### ISSUE TITLE
Artifact Upload Bug: Invalid tarball file is not caught at upload time

### ISSUE THREAD

**Issue by ironfroggy** [ansible/galaxy-dev#247]
##### SUMMARY
Invalid tarball archives submitted to the artifact upload endpoint are not detected and reported as an error in the HTTP response. The tarball only fails later during import. The V2 collection API on Galaxy intentionally detects invalid tarballs early. The V3 API should guarantee the same checks.

##### STEPS TO REPRODUCE
- Create a non-tarball file with a valid artifact filename
- Upload the file to Automation Hub

##### EXPECTED RESULTS
- A 400 response and an error message about not being a valid tar archive

##### ACTUAL RESULTS
A 200 response and an import task endpoint

```

```

**Comment by romanblanco** [comment]
> The V2 collection API on Galaxy intentionally detects invalid tarballs early.

https://github.com/ansible/galaxy/blob/0db1d8105dbf3b9585a84205edb9234c4d10d630/galaxy/worker/tasks/collection.py#L80-L88

**Comment by romanblanco** [comment]
Tracking the progress in branch [issue247](https://github.com/ansible/galaxy-api/compare/master...romanblanco:issue247).

**Comment by romanblanco** [comment]
I've [tried](https://github.com/ansible/galaxy-api/compare/master...romanblanco:issue247#diff-4197903b42a52e1f7f468e56e20a25e8R167-R168) to copy the validation from the API used in [galaxy](https://github.com/ansible/galaxy/blob/devel/galaxy/api/v2/views/collection.py#L126). However, due to our Django configuration, I can't access the uploaded file using `tarfile` library as the uploaded file is serialized as `InMemoryUploadedFile`, not acceptable by `tarfile.is_tarfile` method:

```
data: OrderedDict([('file', <InMemoryUploadedFile: rblanco-collection_demo-1.0.20.tar.gz (ap
plication/gzip)>), ('sha256', None), ('filename', CollectionFilename(namespace='rblanco', name='collection_
demo', version='1.0.20')), ('mimetype', 'application/x-tar')])
```


My options to proceed, suggested in [Django ticket 28402](https://code.djangoproject.com/ticket/28402#comment:4) would be to either to use the bytes of the processed file to write a new file on filesystem that I can later test using `tarfile` or change `FILE_UPLOAD_HANDLERS` configuration.

@cutwater do you have a preferred way to handle this?

**Comment by alikins** [comment]
@romanblanco Is the issue that InMemoryUploadedFile doesn't have a filename / temporary_file_path and that tarfile generally wants a file path?

Maybe it can use the 'fileobj' option for tarfile.open https://docs.python.org/3/library/tarfile.html#tarfile.open ?

**Comment by romanblanco** [comment]
@alikins

> Is the issue that InMemoryUploadedFile doesn't have a filename / temporary_file_path and that tarfile generally wants a file path?

Yeah, it expects _"str, bytes or os.PathLike object, not InMemoryUploadedFile"_

> Maybe it can use the 'fileobj' option for tarfile.open https://docs.python.org/3/library/tarfile.html#tarfile.open ?

Trying that now, it's still not correct: https://github.com/ansible/galaxy-api/commit/4bdf17ac9d7960c0ff9dbe35388b35d5044c3ad8#diff-4197903b42a52e1f7f468e56e20a25e8R186-R193
